build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java

```diff
@@ -60,46 +61,39 @@ public void apply(Project project) {
         project.getPluginManager().apply(ElasticsearchJavaBasePlugin.class);
         var javaExtension = project.getExtensions().getByType(JavaPluginExtension.class);
 
-        var srcDir = project.getProjectDir().toPath().resolve("src");
-        List<Integer> mainVersions = new ArrayList<>();
-        try (var subdirStream = Files.list(srcDir)) {
-            for (Path sourceset : subdirStream.toList()) {
-                assert Files.isDirectory(sourceset);
-                String sourcesetName = sourceset.getFileName().toString();
-                Matcher sourcesetMatcher = MRJAR_SOURCESET_PATTERN.matcher(sourcesetName);
-                if (sourcesetMatcher.matches()) {
-                    mainVersions.add(Integer.parseInt(sourcesetMatcher.group(1)));
-                }
-            }
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
-
-        Collections.sort(mainVersions);
-        List<String> parentSourceSets = new ArrayList<>();
-        parentSourceSets.add(SourceSet.MAIN_SOURCE_SET_NAME);
+        List<Integer> mainVersions = findSourceVersions(project, MRJAR_MAIN_SOURCESET_PATTERN);
+        List<String> mainSourceSets = new ArrayList<>();
+        mainSourceSets.add(SourceSet.MAIN_SOURCE_SET_NAME);
         for (int javaVersion : mainVersions) {
-            String sourcesetName = "main" + javaVersion;
-            addMrjarSourceset(project, javaExtension, sourcesetName, parentSourceSets, javaVersion);
-            parentSourceSets.add(sourcesetName);
+            String sourcesetName = SourceSet.MAIN_SOURCE_SET_NAME + javaVersion;
+            addSourceset(project, javaExtension, sourcesetName, mainSourceSets, javaVersion);
+            mainSourceSets.add(sourcesetName);
         }
-    }
 
-    private void addMrjarSourceset(
-        Project project,
-        JavaPluginExtension javaExtension,
-        String sourcesetName,
-        List<String> parentSourceSets,
-        int javaVersion
-    ) {
-        SourceSet sourceSet = javaExtension.getSourceSets().maybeCreate(sourcesetName);
-        for (String parentSourceSetName : parentSourceSets) {
-            GradleUtils.extendSourceSet(project, parentSourceSetName, sourcesetName);
+        List<Integer> testVersions = findSourceVersions(project, MRJAR_TEST_SOURCESET_PATTERN);
```

Comment:
I don't think we need to do this twice. We can assume for every main source set, there is a corresponding test source set, and if it's empty that's fine the tasks will just get skipped with `NO_SOURCE`.

Reply:
Ah ok, didn't think about NO_SOURCE.